File: app/backend/schemas/schedule.py
from typing import Optional, Literal, List
from pydantic import BaseModel, Field

LocationValue = Literal["telemunca", "birou"]  # simple dropdown

class HourWindow(BaseModel):
    # Day-of-week short name (Mon..Sun) OR a specific date "YYYY-MM-DD"
    day: str
    start: str  # "HH:MM"
    end: str    # "HH:MM"
    action: Literal["exclude", "include"] = "exclude"

class DayConfig(BaseModel):
    enabled: bool = False
    checkIn: Optional[str] = None   # "HH:MM"
    checkOut: Optional[str] = None  # "HH:MM"
    location: Optional[LocationValue] = None

class WeekConfig(BaseModel):
    Mon: DayConfig = DayConfig()
    Tue: DayConfig = DayConfig()
    Wed: DayConfig = DayConfig()
    Thu: DayConfig = DayConfig()
    Fri: DayConfig = DayConfig()
    Sat: DayConfig = DayConfig()
    Sun: DayConfig = DayConfig()

class Jitter(BaseModel):
    minutesMinus: int = 0
    minutesPlus: int = 0

class Holidays(BaseModel):
    publicCalendars: List[str] = Field(default_factory=list)  # e.g., ["RO"]
    personalDates: List[str] = Field(default_factory=list)    # ISO dates
    behavior: Literal["skip", "move_to_next_workday", "run_anyway"] = "skip"

class ExceptionsCfg(BaseModel):
    include: List[str] = Field(default_factory=list)  # ISO "YYYY-MM-DD" or "YYYY-MM-DDTHH:MM"
    exclude: List[str] = Field(default_factory=list)  # same formats; exclude wins over include
    hourWindows: List[HourWindow] = Field(default_factory=list)  # per-day or per-date time windows

class ScheduleSpec(BaseModel):
    tz: str = "Europe/Bucharest"
    week: WeekConfig = WeekConfig()
    jitter: Jitter = Jitter()
    holidays: Holidays = Holidays()
    exceptions: ExceptionsCfg = ExceptionsCfg()

class NextEvent(BaseModel):
    type: Literal["checkIn", "checkOut"]
    at: str        # UTC ISO timestamp
    localDate: str # YYYY-MM-DD
    location: Optional[str] = None

class ScheduleDoc(BaseModel):
    uid: str
    name: str
    active: bool = True
    spec: ScheduleSpec
    next_event: Optional[NextEvent] = None

class RunDoc(BaseModel):
    scheduleId: str
    uid: str
    eventType: Literal["checkIn", "checkOut"]
    scheduledAt: str
    status: Literal["success", "failure", "skipped"] = "success"
    message: Optional[str] = None
    location: Optional[LocationValue] = None

File: web/src/types/schedule.ts
export type Location = "telemunca" | "birou";

export interface DayConfig {
  enabled: boolean;
  checkIn?: string; // "HH:MM"
  checkOut?: string; // "HH:MM"
  location?: Location;
}

export interface WeekConfig {
  Mon: DayConfig;
  Tue: DayConfig;
  Wed: DayConfig;
  Thu: DayConfig;
  Fri: DayConfig;
  Sat: DayConfig;
  Sun: DayConfig;
}

export interface Jitter {
  minutesMinus: number;
  minutesPlus: number;
}

export interface Holidays {
  publicCalendars: string[]; // e.g. ["RO"]
  personalDates: string[]; // "YYYY-MM-DD"
  behavior: "skip" | "move_to_next_workday" | "run_anyway";
}

export interface HourWindow {
  day: string; // "Mon".."Sun" or "YYYY-MM-DD"
  start: string; // "HH:MM"
  end: string; // "HH:MM"
  action?: "exclude" | "include";
}

export interface ExceptionsCfg {
  include: string[]; // ISO dates or datetimes
  exclude: string[]; // ISO dates or datetimes
  hourWindows: HourWindow[];
}

export interface ScheduleSpec {
  tz: string; // e.g. "Europe/Bucharest"
  week: WeekConfig;
  jitter: Jitter;
  holidays: Holidays;
  exceptions: ExceptionsCfg;
}

export interface NextEvent {
  type: "checkIn" | "checkOut";
  at: string; // UTC ISO
  localDate: string; // "YYYY-MM-DD"
  location?: Location;
}

export interface ScheduleDoc {
  id?: string; // added client-side
  uid?: string;
  name: string;
  active: boolean;
  spec: ScheduleSpec;
  next_event?: NextEvent;
}

export interface PreviewEvent {
  type: "checkIn" | "checkOut";
  at: string; // UTC ISO
  localDate: string; // "YYYY-MM-DD"
  location?: Location;
}

export interface PreviewResponse {
  events: PreviewEvent[];
}
